### PR TITLE
*^ Blade CLI update/fix to be operational again

### PR DESCRIPTION
**Pull request by radonnachie**
Outstanding Issues to resolve:

- [ ] segfault when Channel-Step is not all channels (originates from Phasor module, which needs to be able to offset the channel index too)
- [ ] python bind ArrayFrequencyTensor

### CODE REVIEW

File: include/blade/modules/guppi/reader.hh [hunk 1]
```diff
@@ -76,6 +76,48 @@ class BLADE_API Reader : public Module {
...
+
+    const U64 getTotalNumberOfPolarizations() const {
+        return this->getDatashape()->n_pol;
+    }
+
+    const U64 getTotalOutputBufferSize() {
+        return this->getStepOutputBufferDims().size();
+    }
+
+
```
Comment (luigifcruz) [inline include/blade/modules/guppi/reader.hh:120]: Why one method for each dimension value as opposed to `getTotalOutputBufferDims` and `getStepOutputBufferDims`? This breaks the "convention" used by other modules (one method gets'em all).

File: include/blade/modules/bfr5/reader.hh [hunk 2]
```diff
@@ -40,16 +40,32 @@ class BLADE_API Reader : public Module {
...
+    }
+
+    constexpr const U64 getTotalNumberOfFrequencyChannels() const {
+        return this->bfr5.dim_info.nchan;
+    }
+
+    constexpr const U64 getTotalNumberOfPolarizations() const {
+        return this->bfr5.dim_info.npol;
+    }
+
```
Comment (luigifcruz) [inline include/blade/modules/bfr5/reader.hh:68]: Same as https://github.com/luigifcruz/blade/pull/35/files#r976764900

File: src/modules/phasor/generic.cc [hunk 3]
```diff
@@ -33,21 +33,22 @@ Generic<OT>::Generic(const Config& config, const Input& input)
     const F64& min_value = max_value * -1.0;
 
     F64 max_cal = 0.0, min_cal = 0.0;
-    for (const auto& calibration : config.antennaCalibrations) {
-        if (calibration.real() > max_cal) {
-            max_cal = calibration.real();
+    for (size_t i = 0; i < config.antennaCalibrations.size(); i++)
```
Comment (luigifcruz) [inline src/modules/phasor/generic.cc:36]: Style error.

Comment (luigifcruz) [inline src/modules/phasor/generic.cc:36]: Just curious, the `for (const auto& calibration...` did not work here? I can't think why it shouldn't work. Maybe it's a but in the Vector implementation that can be fixed.

File: include/blade/memory/devices/cpu/vector.hh [hunk 4]
```diff
@@ -15,6 +15,8 @@ class BLADE_API Vector<Device::CPU, T, Dims> : public VectorImpl<T, Dims> {
         BL_CHECK_THROW(this->resize(dims));
     }
 
+    explicit Vector(const Vector<Device::CPU, T, Dims>& other) : VectorImpl<T, Dims>(other.data(), other.dims()) {}
```
Comment (luigifcruz) [inline include/blade/memory/devices/cpu/vector.hh:18]: This can be dangerous, is it really necessary?

Comment (radonnachie) [reply include/blade/memory/devices/cpu/vector.hh:18]: I agree, but I'd like to double check how necessary just before merging...

File: include/blade/modules/guppi/reader.hh (hunk 1 above)
Comment (radonnachie) [reply include/blade/modules/guppi/reader.hh:120]: Will clean up.

File: src/modules/phasor/generic.cc (hunk 3 above)
Comment (radonnachie) [reply src/modules/phasor/generic.cc:36]: Will double check

Comment (radonnachie) [reply src/modules/phasor/generic.cc:36]: Will double check.

**Comment by radonnachie** [comment]
Running with:
`apps/blade-cli/blade-cli -t ATA -m B --input-type CI8 --output-type CF16 -c 1 -C 128 -T 128 /mnt/buf0/bladetest_signal_0mean /mnt/buf0/bladetest_signal.bfr5 /mnt/buf0/bladetest_signal_0mean_output`

Specifying no upchannelization, all coarse-channels and a minimum number of time samples.

**Comment by luigifcruz** [comment]
This looks good to me. There are some changes necessary before going to upstream, but I will do them in the target branch. Thanks!

**Pull request merged**